### ISSUE TITLE
Make it easier to detect the end of a media source

### ISSUE THREAD

**Issue by dbeckwith** [pdeljanov/Symphonia#134]
As a suggestion, it would be nice if either the API or documentation made it more obvious how you can detect the end of a media source. Currently, I do something like this:
```rust
match format_reader.next_packet() {
    Ok(packet) => {
        // got the next packet
    },
    Err(symphonia::core::errors::Error::IoError(error))
        if error.kind() == std::io::ErrorKind::UnexpectedEof =>
    {
        // audio source ended
    }
    Err(error) => {
        // some other unrecoverable error occurred 
    },
}
```
But it seems like a lot of boilerplate and not very discoverable or obvious. I'm also not 100% sure `UnexpectedEof` is the only error that should be treated as end-of-stream.

It would be nice if `next_packet` could return a `None` when there are no more packets in the stream, but that would be a breaking change (could maybe add a new method like `next_packet_opt` or something). Or, at least the documentation for `next_packet` should describe how to inspect the errors to handle this.

**Comment by qarmin** [comment]
Related issue and explanation https://github.com/pdeljanov/Symphonia/discussions/121#discussioncomment-2711681

**Comment by pdeljanov** [comment]
This seems to be a question/criticism that seems to come up a lot.

Since it's coming up so frequently I will try to address it in a future release. However, as mentioned, this would be a breaking change so it may be a while before it lands.

I think there are two possible options:

1. As suggested, return `None` from `next_packet` for an iterator-like interface.
2. Return an `EndOfStream` or `EndOfMedia` error.

In both of these cases, an `UnexpectedEof` error would go back to truly being an unexpected end-of-file.

Any thoughts? The former is probably more Rust-like, but the latter avoids returning a `Result<Option<Packet>>`

**Comment by dbeckwith** [comment]
I think I like option 1 better because it forces you handle the end of the stream using types, whereas with an error variant it's much easier to just propagate the error and you're less likely to consider the stream ending. But if you wanted to avoid a breaking change, you could do option 2 for now. Then if you're planning a bunch of breaking changes in the future you could maybe then both remove the `EndOfStream` error and finally go with option 1.

**Comment by DoumanAsh** [comment]
> Any thoughts? The former is probably more Rust-like, but the latter avoids returning a Result<Option<Packet>>
Just return empty packet with zero sized data and document this behavior

**Comment by pdeljanov** [comment]
This has now been implemented in the `dev-0.6` branch!

**Comment by ahicks92** [comment]
I came back to my projects using Symphonia.

Any chance that 0.6 will happen soon?

Without this I can't tell for sure (or at least I don't think I can tell for sure) whether failures to decode are my bug, Symphonia's bug, or just eof.

It looks like this project slowed down a lot, so maybe you could do a 0.6 with just this change?  I mean I acknowledge that might also be really hard so if not that's all right, but there seems little harm in asking.
